### ISSUE TITLE
[NLP] Add option for using the pooling layer in Text Embedding models

### ISSUE THREAD

**Issue by davidkyle** [elastic/eland#531]
BERT based text embedding models usually output a pooling layer and the last hidden state. The first element of the last hidden state is the CLS token which is the embedding representation the entire input. Conventionally the last hidden state is the first output of the model and this is what is returned by the C++ code.

https://github.com/elastic/ml-cpp/blob/main/bin/pytorch_inference/Main.cc#L63

The Java then takes the first vector from the tensor which is the CLS token

https://github.com/elastic/elasticsearch/blob/a8d57233093385169d158e931995a50bf9a135aa/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/TextEmbeddingProcessor.java#LL57C54-L57C54

It takes quite a deep dive to figure out that ES is returning the CLS token and not the output of some pooling layer. There is some debate as to which is best to use, some might prefer to use a pooling layer and others the CLS token.

To avoid painful code changes in C++ and Java it would be easiest to wrap the model to select preferred output in Eland but making it the first item returned.

Note: Eland removes the pooling layer from Sentence Transformer models 
https://github.com/elastic/eland/blob/main/eland/ml/pytorch/transformers.py#L306

**Comment by davidkyle** [comment]
This discuss post has sample code to wrap a Sentence Transformer model and return the mean pooling output

https://discuss.elastic.co/t/text-embedding-different-in-elastic-search-and-python-library/329164/5

**Comment by temiwale88** [comment]
Thanks @davidkyle . I'm happy I searched well. This is exactly what I need. I will try out the 2nd option: editing [transformers.py](https://github.com/elastic/eland/blob/main/eland/ml/pytorch/transformers.py).

**Before I do this** please confirm that `_SentenceTransformerWrapperModule` in transformer.py doesn't already do this? 

Here are the comments:
    """
    _A wrapper around sentence-transformer models to provide pooling,
    normalization and other graph layers that are not defined in the base
    HuggingFace transformer model._
    """

**Comment by temiwale88** [comment]
@davidkyle . I just wanted to add that before you output the model embeddings as is, you might need to normalize it first and output that value instead e.g.

```python
import torch.nn.functional as F
model_output = F.normalize(embedding, p=2, dim=1)
```

**Comment by joshdevins** [comment]
@davidkyle Are you sure about this? The discuss thread mentions that the outputs are different but they are not, Python has just formatted the string differently. Everything should work OOTB for sentence-transformers. ES isn't returning the CLS token but actually the mean pooled embedding from sentence-transformers (IIRC). I can spend some time debugging this if it's a real problem but I'm not sure it is. Any more context would be good as this seems to be a major issue if it's a real bug.
